File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    let n = preorder.length
    return helper(preorder,0,n-1,inorder,0,n-1)
    
    
};

function helper(preorder:number[],prelow:number,prehigh:number,inorder:number[],inlow:number,inhigh:number){
    if (prelow > prehigh) return null
    let root = new TreeNode(preorder[prelow])
    let i = inlow
    while(preorder[prelow]!=inorder[i]) i++;
    let leftSize = i - inlow
    root.left = helper(preorder,prelow+1,prelow+leftSize,inorder,inlow,i-1)
    root.right = helper(preorder,prelow+leftSize+1,prehigh,inorder,i+1,inhigh)
    return root
    
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer,Integer> inorderMap = new HashMap<>();
        for (int i = 0; i < preorder.length; i++){
            inorderMap.put(inorder[i],i);
        }
        return helper(0,preorder.length-1,0,preorder,inorderMap);
    }

    public TreeNode helper(int left,int right,int idx,int[] preorder,Map<Integer,Integer> inorderMap){
    TreeNode node = new TreeNode(preorder[idx]);
    int mid = inorderMap.get(preorder[idx]);
    if (mid > left) node.left = helper(left,mid-1,idx+1,preorder,inorderMap);
    if (mid < right) node.right = helper(mid+1,right,mid+idx-left+1,preorder,inorderMap);
    return node;
    }
}